[nri-prometheus] Fix nri-prometheus-config to include config.transformation content in lowDataMode.

shawnj opened the pull request:
<!--
Thank you for contributing to New Relic's Helm charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements:

* https://github.com/newrelic-experimental/helm-charts/blob/master/CONTRIBUTING.md#technical-requirements

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/newrelic-experimental/helm-charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a Github Action
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->
#### Is this a new chart
No

#### What this PR does / why we need it:
This fix merges the `transformations` content from the `lowDataDefault` yaml with the `config.transformations` content from the default `.Values.config.transformations` content in the `values.yaml` file.

Current configuration overwrites the `.Values.config.transformation` content with the content from the `lowDataDefault` yaml content when `lowDataMode` is `true`.

#### Notes for reviewer:
Needed to use a "manual" approach for merging `transformation` content as the helm `merge` function will not merge different object types.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[mychartname]`)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/newrelic/helm-charts?pullRequest=643) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/newrelic/helm-charts?pullRequest=643) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/newrelic/helm-charts?pullRequest=643) it.</sub>

kang-makes commented on charts/nri-prometheus/templates/configmap.yaml:
```diff
@@ -9,10 +9,17 @@ data:
-    {{- mergeOverwrite .Values.config $lowDataDefault | toYaml | indent 4 -}}
-{{- else }}
-    {{ toYaml .Values.config | indent 4 | trim }}
+    {{ omit .Values.config "transformations" | toYaml | indent 4 | trim -}}
```
I would `trim` before `indent` and use `{{-` with `nindent` so it will clean all the YAML and do the indentantion properly an consistently.

kang-makes commented on charts/nri-prometheus/templates/configmap.yaml:
```diff
@@ -9,10 +9,17 @@ data:
+    transformations: 
+    {{ .Values.config.transformations | toYaml | indent 4 | trim -}}
+{{- end }}
+{{- if (include "nri-prometheus.lowDataMode" .) -}}
+    {{ $lowDataDefault := .Files.Get "static/lowdatamodedefaults.yaml" | fromYaml }}
+    {{ if .Values.config.transformations }}
+    {{- $lowDataDefault.transformations | toYaml | indent 4 | trim }}
+    {{ else }}
+    {{- $lowDataDefault | toYaml | indent 4 | trim }}
+    {{ end }}
```
This could lead to two `transformations` in case of `transformations` in the config and `lowDataMode` enabled.

A `merge` with the defaults (in case of `lowDataMode`) with the one that came from `.Values.configtransformations` should be needed or a way to allow overriding `transformations` properly.

kang-makes reviewed (changes requested):
There is a possibility that `transformations` could be declared twice making an invalid configuration file. Could I read the template code incorrectly?

kang-makes commented:
Hhmm... Maybe it is correct and I am reading it incorrectly but a helper could be awesome so we have this obscure logic more readable...

Let me recheck it.

shawnj commented:
@kang-makes I've added a `mergeTransformations` function to `helpers` to hopefully clear some of the confusion.  Also, I tested the various configurations and since I'm omitting `transformations` from the original config and replacing it there are no duplicate `transformation` sections in the config.

I've also remove the `trim` and cleaned up the whitespace a bit.

kang-makes commented:
I would like to merge this by myself but I cannot because of a couple of reasons:
- Next time, instead for merging from our `master` branch is better to `rebase` because we can replay the commits easier and I can take ownership or you branch maintaining your authorship of the commits.
- Instead of using you `master` branch on your fork, use other branch to make the rebase easier too.

I would like to take ownership of this and bump the chart version because the `master` version has gone forward this PR but I cannot because of these issues.

Can you bump the version so I can merge it? It would be a shame to lose all this work already done.

shawnj commented:
@kang-makes Appreciate the guidance on branching for future PR's.  I have bumped the fix version for the `nri-prometheus` chart.

kang-makes reviewed (approved):
LGTM
